SI-6536 Generates super accessors X.super[Y].blah when Y is a class

JamesIry opened the pull request:
The main change here is to add another case for generating super accessors -
the case in X.super[Y].blah when X isn't the current class and Y is a class.
The change is deliberately kept as minimal as possible to reduce the chance
of breaking something in the 2.9.x line.

Additionally GenICode now detects the case when we're trying to emit
byte code that would be nonsense and warns about it. That
can safely be made an assert for 2.11.

Finally a related assert in RefChecks is beefed up to output a bit more useful
information.

review @paulp, @adriaanm

Commit:
SI-6536 Generates super accessors X.super[Y].blah when Y is a class

The main change here is to add another case for generating super accessors -
the case in X.super[Y].blah when X isn't the current class and Y is a class.
The change is deliberately kept as minimal as possible to reduce the chance
of breaking something in the 2.9.x line.

Additionally GenICode now detects the case when we're trying to emit
byte code that would be nonsense and warns about it. That
can safely be made an assert for 2.11.

Finally a related assert in RefChecks is beefed up to output a bit more useful
information.

paulp commented on src/compiler/scala/tools/nsc/backend/icode/GenICode.scala:
```diff
@@ -767,11 +767,15 @@ abstract class GenICode extends SubComponent  {
+        case Apply(fun @ Select(Super(qual, mix), _), args) =>
+
+          if (!qual.isInstanceOf[This]) {
+            log("WARNING: super call where selector isn't 'this'. May generate invalid bytecode. Previous phases should have tansformed away this form of super.")
```
Where I come from the word is "transform". NEWB

JamesIry replied:
No, tansform. It's Californian.

paulp commented on src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala:
```diff
@@ -118,7 +119,20 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
+      // SI-6536 has more discussion about how this works.
+      if (name.isTermName && (mix == tpnme.EMPTY && (clazz.isTrait || clazz != currentClass || !validCurrentOwner ))
+                          || (mix != tpnme.EMPTY && !mixtpe.typeSymbol.isTrait && clazz != currentClass))
         ensureAccessor(sel)
```
Can you please extract a condition of this length into a local method? This has no binary compatibility implications (it's private) and is far easier to  read and maintain. I would maybe write it like this:

```
def requiresAccessor = name.isTermName && (clazz != currentClass) && (mix match {
  case tpnme.EMPTY => clazz.isTrait || !validCurrentOwner
  case _           => !mixtpe.typeSymbol.isTrait
})
```

Ha ha, "mix match".

paulp replied:
I borked the logic of the above, but the gist is stil sound (especially the 2-case match on mix)

JamesIry commented on src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala:
```diff
@@ -110,6 +110,7 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
       val Select(sup @ Super(_, mix), name) = sel
       val sym   = sel.symbol
       val clazz = sup.symbol
+      val SuperType(_, mixtpe) = sup.tpe
```
Should have a better error if the assumption that sup.tpe is a SuperType turns out to be wrong.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1168/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1168/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1883/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1883/

Commit:
SI-6536 Cleanup code around determining accessor requirement

Moves the logic for the condition on requiring an accessor into a local method
with a match to make thing cleaner. Also determines the mix type using
a pattern match that logs but does not fail if the symbol isn't of the expected
type. Finally fixes a typo in an assertion in GenICode.

review @paulp

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1171/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1171/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1886/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1886/

adriaanm commented:
LGTM
